javatests/arcs/core/analysis/ExpressionDataFlowAnalyzerTest.kt

```diff
@@ -0,0 +1,370 @@
+  }
+
+  @Test
+  fun sum_variable_literal() {
+    val expr = PaxelParser.parse("x + 1")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.DerivedFrom(DependencyNode.PrimitiveValue("x")))
+  }
+
+  @Test
+  fun variable_field_access() {
+    val expr = PaxelParser.parse("x.foo")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.PrimitiveValue("x", "foo"))
+  }
+
+  @Test
+  fun variable_field_access_nested() {
+    val expr = PaxelParser.parse("x.foo.bar")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.PrimitiveValue("x", "foo", "bar"))
+  }
+
+  @Test
+  fun variable_field_access_binop() {
+    val expr = PaxelParser.parse("x.foo.bar + y.foo.bar.baz + z.baz.bar")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(
+      DependencyNode.DerivedFrom(
+        DependencyNode.PrimitiveValue("x", "foo", "bar"),
+        DependencyNode.PrimitiveValue("y", "foo", "bar", "baz"),
+        DependencyNode.PrimitiveValue("z", "baz", "bar")
+      )
+    )
+  }
+
+  @Test
+  fun literal_binop() {
+    val expr = PaxelParser.parse("1 + 2 + 3 + 4 + 5")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.LITERAL)
+  }
+
+  @Test
+  fun variable_field_access_binop_literals() {
+    val expr = PaxelParser.parse("x.foo.bar + y.foo.bar.baz + z.baz.bar + 10 + 20")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(
+      DependencyNode.DerivedFrom(
+        DependencyNode.PrimitiveValue("x", "foo", "bar"),
+        DependencyNode.PrimitiveValue("y", "foo", "bar", "baz"),
+        DependencyNode.PrimitiveValue("z", "baz", "bar")
+      )
+    )
+  }
+
+  @Test
+  fun new() {
+    val expr = PaxelParser.parse("new Object {foo: x, bar: y}")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(
+      DependencyNode.AggregateValue(
+        "foo" to DependencyNode.PrimitiveValue("x"),
+        "bar" to DependencyNode.PrimitiveValue("y")
+      )
+    )
+  }
+
+  @Test
+  fun new_field_access() {
+    val expr = PaxelParser.parse("new Object {foo: input.foo, bar: input.foo.bar}")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(
+      DependencyNode.AggregateValue(
+        "foo" to DependencyNode.PrimitiveValue("input", "foo"),
+        "bar" to DependencyNode.PrimitiveValue("input", "foo", "bar")
+      )
+    )
+  }
+
+  @Test
+  fun sum_variables() {
+    val expr = PaxelParser.parse("x + y")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(
+      DependencyNode.DerivedFrom(
+        DependencyNode.PrimitiveValue("x"),
+        DependencyNode.PrimitiveValue("y")
+      )
+    )
+  }
+
+  @Test
+  fun product_same_variable() {
+    val expr = PaxelParser.parse("x * x")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.DerivedFrom(listOf("x")))
```

Comment:
I find it slightly surprising that you can create a `DependencyNode.DerivedFrom` from either a list of string or list of DependencyNodes. I do understand that we interpret one as a path and as a utility, but I think it's making the API more confusing that it's helping. I would recommend removing the `constructor(vararg paths: Path) : this(paths.map { PrimitiveValue(it) }.toSet())` constructor.